README.md
# cmp_polymorphism

![Rust Version][rustc-image]
![Apache2/MIT licensed][license-image]

research: compare polymorphism

## Benchmark Results

rustc 1.56.1 (59eed8a2a 2021-11-01)
|         `name`         |   `bench`   |  `per 1k`   |
|:-----------------------|------------:|------------:|
| enum_obj^              |    0.025 kc |    0.000 kc |
| enum_obj^vec^id^01k    |    4.576 kc |    4.576 kc |
| enum_obj^vec^id^08k    |   36.355 kc |    4.544 kc |
| enum_obj^vec^id^90k    | 1270.354 kc |   14.115 kc |
| enum_obj^vec^rem^01k   |   12.587 kc |   12.587 kc |
| enum_obj^vec^rem^08k   |   99.855 kc |   12.482 kc |
| enum_obj^vec^rem^90k   | 1914.675 kc |   21.274 kc |
| enum_obj^vec^sum^01k   |    7.910 kc |    7.910 kc |
| enum_obj^vec^sum^08k   |   87.262 kc |   10.908 kc |
| enum_obj^vec^sum^90k   | 1597.248 kc |   17.747 kc |
| trait_obj^             |    0.025 kc |    0.000 kc |
| trait_obj^vec^id^01k   |    6.858 kc |    6.858 kc |
| trait_obj^vec^id^08k   |   49.307 kc |    6.163 kc |
| trait_obj^vec^id^90k   |  627.355 kc |    6.971 kc |
| trait_obj^vec^rem^01k  |   10.774 kc |   10.774 kc |
| trait_obj^vec^rem^08k  |   84.752 kc |   10.594 kc |
| trait_obj^vec^rem^90k  | 2287.182 kc |   25.413 kc |
| trait_obj^vec^sum^01k  |    8.720 kc |    8.720 kc |
| trait_obj^vec^sum^08k  |   71.800 kc |    8.975 kc |
| trait_obj^vec^sum^90k  | 2374.904 kc |   26.388 kc |

rustc 1.53.0 (53cb7b09b 2021-06-17)
|         `name`         |   `bench`   |  `per 1k`   |
|:-----------------------|------------:|------------:|
| enum_obj^              |    0.024 kc |    0.000 kc |
| enum_obj^vec^id^01k    |    4.358 kc |    4.358 kc |
| enum_obj^vec^id^08k    |   35.195 kc |    4.399 kc |
| enum_obj^vec^id^90k    |  604.227 kc |    6.714 kc |
| enum_obj^vec^rem^01k   |   11.488 kc |   11.488 kc |
| enum_obj^vec^rem^08k   |   91.363 kc |   11.420 kc |
| enum_obj^vec^rem^90k   | 1198.158 kc |   13.313 kc |
| enum_obj^vec^sum^01k   |   10.501 kc |   10.501 kc |
| enum_obj^vec^sum^08k   |   60.464 kc |    7.558 kc |
| enum_obj^vec^sum^90k   |  930.674 kc |   10.341 kc |
| trait_obj^             |    0.024 kc |    0.000 kc |
| trait_obj^vec^id^01k   |    7.092 kc |    7.092 kc |
| trait_obj^vec^id^08k   |   56.413 kc |    7.052 kc |
| trait_obj^vec^id^90k   |  637.158 kc |    7.080 kc |
| trait_obj^vec^rem^01k  |   10.077 kc |   10.077 kc |
| trait_obj^vec^rem^08k  |   80.625 kc |   10.078 kc |
| trait_obj^vec^rem^90k  | 1364.740 kc |   15.164 kc |
| trait_obj^vec^sum^01k  |    8.369 kc |    8.369 kc |
| trait_obj^vec^sum^08k  |   66.302 kc |    8.288 kc |
| trait_obj^vec^sum^90k  | 1203.361 kc |   13.371 kc |

rustc 1.52.0 (88f19c6da 2021-05-03)
|         `name`         |   `bench`   |  `per 1k`   |
|:-----------------------|------------:|------------:|
| enum_obj^              |    0.024 kc |    0.000 kc |
| enum_obj^vec^id^01k    |    4.425 kc |    4.425 kc |
| enum_obj^vec^id^08k    |   35.634 kc |    4.454 kc |
| enum_obj^vec^id^90k    |  583.796 kc |    6.487 kc |
| enum_obj^vec^rem^01k   |   11.453 kc |   11.453 kc |
| enum_obj^vec^rem^08k   |   90.914 kc |   11.364 kc |
| enum_obj^vec^rem^90k   | 1182.439 kc |   13.138 kc |
| enum_obj^vec^sum^01k   |   10.670 kc |   10.670 kc |
| enum_obj^vec^sum^08k   |   79.980 kc |    9.998 kc |
| enum_obj^vec^sum^90k   |  900.310 kc |   10.003 kc |
| trait_obj^             |    0.024 kc |    0.000 kc |
| trait_obj^vec^id^01k   |    6.062 kc |    6.062 kc |
| trait_obj^vec^id^08k   |   48.345 kc |    6.043 kc |
| trait_obj^vec^id^90k   |  546.009 kc |    6.067 kc |
| trait_obj^vec^rem^01k  |   10.162 kc |   10.162 kc |
| trait_obj^vec^rem^08k  |   80.179 kc |   10.022 kc |
| trait_obj^vec^rem^90k  | 1280.631 kc |   14.229 kc |
| trait_obj^vec^sum^01k  |    8.413 kc |    8.413 kc |
| trait_obj^vec^sum^08k  |   66.099 kc |    8.262 kc |
| trait_obj^vec^sum^90k  | 1182.313 kc |   13.137 kc |


[//]: # (badges)

[rustc-image]: https://img.shields.io/badge/rustc-1.56+-blue.svg
[license-image]: https://img.shields.io/badge/license-Apache2.0/MIT-blue.svg


CHANGELOG.md
# Changelog: cmp_polymorphism

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] *
### Added
* badges into `README.md`

### Changed
* reformat `CHANGELOG.md`


## [0.1.4] (2021-11-04)
### Changed
* update depends

## [0.1.3] (2021-06-30)
### Added
* memx-cdy(0.1) into `dev-dependencies`

### Changed
* update depends: anyhow(1.0.41)

## [0.1.2] (2021-05-10)
### Changed
* update depends: regex(1.5.4)
* more short measuring and warming time

## [0.1.1] (2021-04-25)
### Changed
* update depends: regex(1.4.6)
* update depends: anyhow(1.0.40): Reduce memory footprint of errors on Rust versions 1.51+
* update depends: many

## [0.1.0] (2021-01-30)
* first commit

[Unreleased]: https://github.com/aki-akaguma/cmp_polymorphism/compare/v0.1.4..HEAD
[0.1.4]: https://github.com/aki-akaguma/cmp_polymorphism/compare/v0.1.3..v0.1.4
[0.1.3]: https://github.com/aki-akaguma/cmp_polymorphism/compare/v0.1.2..v0.1.3
[0.1.2]: https://github.com/aki-akaguma/cmp_polymorphism/compare/v0.1.1..v0.1.2
[0.1.1]: https://github.com/aki-akaguma/cmp_polymorphism/compare/v0.1.0..v0.1.1
[0.1.0]: https://github.com/aki-akaguma/cmp_polymorphism/releases/tag/v0.1.0


src/lib.rs
pub mod enum_obj;
pub mod trait_obj;

pub fn do_trait_obj(
    count: i32,
) -> anyhow::Result<(
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
)> {
    let a: Box<dyn trait_obj::Animal> = Box::new(trait_obj::Cat::new(count));
    let b: Box<dyn trait_obj::Animal> = Box::new(trait_obj::Dog::new(count, count + 1));
    let c: Box<dyn trait_obj::Animal> = Box::new(trait_obj::Duck::new(count, count + 1, count + 2));
    let d: Box<dyn trait_obj::Animal> =
        Box::new(trait_obj::Crow::new(count, count + 1, count + 2, count + 3));
    let e: Box<dyn trait_obj::Animal> = Box::new(trait_obj::Frog::new(
        count,
        count + 1,
        count + 2,
        count + 3,
        count + 4,
    ));
    Ok((
        (a.talk(), a.sum()),
        (b.talk(), b.sum()),
        (c.talk(), c.sum()),
        (d.talk(), d.sum()),
        (e.talk(), e.sum()),
    ))
}

pub fn do_enum_obj(
    count: i32,
) -> anyhow::Result<(
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
)> {
    let a: Box<enum_obj::Animal> = Box::new(enum_obj::Animal::Cat(count));
    let b: Box<enum_obj::Animal> = Box::new(enum_obj::Animal::Dog(count, count + 1));
    let c: Box<enum_obj::Animal> = Box::new(enum_obj::Animal::Duck(count, count + 1, count + 2));
    let d: Box<enum_obj::Animal> = Box::new(enum_obj::Animal::Crow(
        count,
        count + 1,
        count + 2,
        count + 3,
    ));
    let e: Box<enum_obj::Animal> = Box::new(enum_obj::Animal::Frog(
        count,
        count + 1,
        count + 2,
        count + 3,
        count + 4,
    ));
    Ok((
        (a.talk(), a.sum()),
        (b.talk(), b.sum()),
        (c.talk(), c.sum()),
        (d.talk(), d.sum()),
        (e.talk(), e.sum()),
    ))
}

pub fn create_trait_objs(count: usize) -> Vec<Box<dyn trait_obj::Animal>> {
    let v: Vec<Box<dyn trait_obj::Animal>> = vec![
        Box::new(trait_obj::Cat::new(1)) as Box<dyn trait_obj::Animal>,
        Box::new(trait_obj::Dog::new(1, 2)) as Box<dyn trait_obj::Animal>,
        Box::new(trait_obj::Duck::new(1, 2, 3)) as Box<dyn trait_obj::Animal>,
        Box::new(trait_obj::Crow::new(1, 2, 3, 4)) as Box<dyn trait_obj::Animal>,
        Box::new(trait_obj::Frog::new(1, 2, 3, 4, 5)) as Box<dyn trait_obj::Animal>,
    ]
    .into_iter()
    .cycle()
    .take(count)
    .collect();
    v
}
pub fn sum_id_trait_objs(vec: &Vec<Box<dyn trait_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.animal_id();
    }
    acc
}
pub fn sum_sum_trait_objs(vec: &Vec<Box<dyn trait_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.sum();
    }
    acc
}
pub fn sum_rem_trait_objs(vec: &Vec<Box<dyn trait_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.rem();
    }
    acc
}

pub fn create_enum_objs(count: usize) -> Vec<Box<enum_obj::Animal>> {
    let v: Vec<Box<enum_obj::Animal>> = vec![
        Box::new(enum_obj::Animal::Cat(1)) as Box<enum_obj::Animal>,
        Box::new(enum_obj::Animal::Dog(1, 2)) as Box<enum_obj::Animal>,
        Box::new(enum_obj::Animal::Duck(1, 2, 3)) as Box<enum_obj::Animal>,
        Box::new(enum_obj::Animal::Crow(1, 2, 3, 4)) as Box<enum_obj::Animal>,
        Box::new(enum_obj::Animal::Frog(1, 2, 3, 4, 5)) as Box<enum_obj::Animal>,
    ]
    .into_iter()
    .cycle()
    .take(count)
    .collect();
    v
}
pub fn sum_id_enum_objs(vec: &Vec<Box<enum_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.animal_id();
    }
    acc
}
pub fn sum_sum_enum_objs(vec: &Vec<Box<enum_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.sum();
    }
    acc
}
pub fn sum_rem_enum_objs(vec: &Vec<Box<enum_obj::Animal>>) -> i32 {
    let mut acc = 0;
    for a in vec {
        acc += a.rem();
    }
    acc
}


src/enum_obj.rs
#[derive(Clone)]
pub enum Animal {
    Cat(i32),
    Dog(i32, i32),
    Duck(i32, i32, i32),
    Crow(i32, i32, i32, i32),
    Frog(i32, i32, i32, i32, i32),
}
impl Animal {
    pub fn animal_id(&self) -> i32 {
        match self {
            Animal::Cat(_a) => 1,
            Animal::Dog(_a, _b) => 2,
            Animal::Duck(_a, _b, _c) => 3,
            Animal::Crow(_a, _b, _c, _d) => 4,
            Animal::Frog(_a, _b, _c, _d, _e) => 5,
        }
    }
    pub fn talk(&self) -> &'static str {
        match self {
            Animal::Cat(_a) => "Meow!",
            Animal::Dog(_a, _b) => "Woof!",
            Animal::Duck(_a, _b, _c) => "Quack!",
            Animal::Crow(_a, _b, _c, _d) => "Caw!",
            Animal::Frog(_a, _b, _c, _d, _e) => "Croak!",
        }
    }
    pub fn sum(&self) -> i32 {
        match self {
            Animal::Cat(a) => *a,
            Animal::Dog(a, b) => a + b,
            Animal::Duck(a, b, c) => a + b + c,
            Animal::Crow(a, b, c, d) => a + b + c + d,
            Animal::Frog(a, b, c, d, e) => a + b + c + d + e,
        }
    }
    pub fn rem(&self) -> i32 {
        match self {
            Animal::Cat(a) => *a,
            Animal::Dog(a, b) => b % a,
            Animal::Duck(_a, b, c) => c % b,
            Animal::Crow(_a, _b, c, d) => d % c,
            Animal::Frog(_a, _b, _c, d, e) => e % d,
        }
    }
}


src/trait_obj.rs
pub trait Animal {
    fn animal_id(&self) -> i32;
    fn talk(&self) -> &'static str;
    fn sum(&self) -> i32;
    fn rem(&self) -> i32;
    fn box_clone(&self) -> Box<dyn Animal>;
}
impl Clone for Box<dyn Animal> {
    fn clone(&self) -> Box<dyn Animal> {
        self.box_clone()
    }
}

#[derive(Clone)]
pub struct Cat(i32);
#[derive(Clone)]
pub struct Dog(i32, i32);
#[derive(Clone)]
pub struct Duck(i32, i32, i32);
#[derive(Clone)]
pub struct Crow(i32, i32, i32, i32);
#[derive(Clone)]
pub struct Frog(i32, i32, i32, i32, i32);

impl Cat {
    pub fn new(a: i32) -> Self {
        Self(a)
    }
}
impl Dog {
    pub fn new(a: i32, b: i32) -> Self {
        Self(a, b)
    }
}
impl Duck {
    pub fn new(a: i32, b: i32, c: i32) -> Self {
        Self(a, b, c)
    }
}
impl Crow {
    pub fn new(a: i32, b: i32, c: i32, d: i32) -> Self {
        Self(a, b, c, d)
    }
}
impl Frog {
    pub fn new(a: i32, b: i32, c: i32, d: i32, e: i32) -> Self {
        Self(a, b, c, d, e)
    }
}

impl Animal for Cat {
    fn animal_id(&self) -> i32 {
        1
    }
    fn talk(&self) -> &'static str {
        "Meow!"
    }
    fn sum(&self) -> i32 {
        self.0
    }
    fn rem(&self) -> i32 {
        self.0
    }
    fn box_clone(&self) -> Box<dyn Animal> {
        Box::new(self.clone())
    }
}
impl Animal for Dog {
    fn animal_id(&self) -> i32 {
        2
    }
    fn talk(&self) -> &'static str {
        "Woof!"
    }
    fn sum(&self) -> i32 {
        self.0 + self.1
    }
    fn rem(&self) -> i32 {
        self.1 % self.0
    }
    fn box_clone(&self) -> Box<dyn Animal> {
        Box::new(self.clone())
    }
}
impl Animal for Duck {
    fn animal_id(&self) -> i32 {
        3
    }
    fn talk(&self) -> &'static str {
        "Quack!"
    }
    fn sum(&self) -> i32 {
        self.0 + self.1 + self.2
    }
    fn rem(&self) -> i32 {
        self.2 % self.1
    }
    fn box_clone(&self) -> Box<dyn Animal> {
        Box::new(self.clone())
    }
}
impl Animal for Crow {
    fn animal_id(&self) -> i32 {
        4
    }
    fn talk(&self) -> &'static str {
        "Caw!"
    }
    fn sum(&self) -> i32 {
        self.0 + self.1 + self.2 + self.3
    }
    fn rem(&self) -> i32 {
        self.3 % self.2
    }
    fn box_clone(&self) -> Box<dyn Animal> {
        Box::new(self.clone())
    }
}
impl Animal for Frog {
    fn animal_id(&self) -> i32 {
        5
    }
    fn talk(&self) -> &'static str {
        "Croak!"
    }
    fn sum(&self) -> i32 {
        self.0 + self.1 + self.2 + self.3 + self.4
    }
    fn rem(&self) -> i32 {
        self.4 % self.3
    }
    fn box_clone(&self) -> Box<dyn Animal> {
        Box::new(self.clone())
    }
}


benches/bench-enum-obj.rs
use criterion::{criterion_group, criterion_main, Criterion};
use criterion_cycles_per_byte::CyclesPerByte;

fn process_one(
    count: i32,
) -> anyhow::Result<(
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
)> {
    cmp_polymorphism::do_enum_obj(count)
}

fn criterion_benchmark(c: &mut Criterion<CyclesPerByte>) {
    match process_one(criterion::black_box(criterion::black_box(10))) {
        Ok(((ct, cn), (dt, dn), (ut, un), (rt, rn), (gt, gn))) => {
            assert_eq!(ct, "Meow!");
            assert_eq!(cn, 10);
            assert_eq!(dt, "Woof!");
            assert_eq!(dn, 10 + 10 + 1);
            assert_eq!(ut, "Quack!");
            assert_eq!(un, 10 + 10 + 1 + 10 + 2);
            assert_eq!(rt, "Caw!");
            assert_eq!(rn, 10 + 10 + 1 + 10 + 2 + 10 + 3);
            assert_eq!(gt, "Croak!");
            assert_eq!(gn, 10 + 10 + 1 + 10 + 2 + 10 + 3 + 10 + 4);
        }
        Err(err) => {
            eprintln!("{}", err);
            unreachable!();
        }
    }
    //
    c.bench_function("enum_obj^", |b| {
        b.iter(|| {
            let _r = process_one(criterion::black_box(10));
        })
    });
    //
    {
        let vec = cmp_polymorphism::create_enum_objs(criterion::black_box(1000));
        c.bench_function("enum_obj^vec^id^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^sum^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^rem^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_enum_objs(&vec);
            })
        });
    }
    //
    {
        let vec = cmp_polymorphism::create_enum_objs(criterion::black_box(8000));
        c.bench_function("enum_obj^vec^id^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^sum^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^rem^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_enum_objs(&vec);
            })
        });
    }
    //
    {
        let vec = cmp_polymorphism::create_enum_objs(criterion::black_box(90000));
        c.bench_function("enum_obj^vec^id^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^sum^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_enum_objs(&vec);
            })
        });
        c.bench_function("enum_obj^vec^rem^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_enum_objs(&vec);
            })
        });
    }
}

criterion_group!(
    name = benches;
    config = Criterion::default().with_measurement(CyclesPerByte)
        .warm_up_time(std::time::Duration::from_millis(300))
        .measurement_time(std::time::Duration::from_millis(1500));
    targets = criterion_benchmark);
criterion_main!(benches);


xtask/src/shape_benchmark_results.rs
use anyhow::Context;
use std::io::BufRead;

pub fn run(_program: &str, _args: &[&str]) -> anyhow::Result<()> {
    let mut bench_vec = get_bench("z.bench.log")?;
    //set_size(&mut bench_vec, "z.size-release.curl.log")?;
    output(bench_vec)?;
    //
    //let mut bench_vec = get_bench("z.bench-release-s.curl.log")?;
    //set_size(&mut bench_vec, "z.size-release.curl.log")?;
    //output(bench_vec)?;
    //
    Ok(())
}

fn output(bench_vec: Vec<BenchStr>) -> anyhow::Result<()> {
    println!(
        "| {:^22} | {:^11} | {:^11} |",
        "`name`", "`bench`", "`per 1k`"
    );
    println!(
        "|:{:<22}-|-{:>11}:|-{:>11}:|",
        "-".repeat(22),
        "-".repeat(11),
        "-".repeat(11),
    );
    for bench in bench_vec {
        if bench.is_cycle {
            println!(
                "| {:<22} | {:>8.3} kc | {:>8.3} kc |",
                bench.name,
                bench.time / 1000.0,
                bench.time_1k / 1000.0,
            );
        } else {
            println!(
                "| {:<22} | {:>8.3} uc | {:>8.3} uc |",
                bench.name,
                bench.time / 0.000001,
                bench.time_1k / 0.000001,
            );
        }
    }
    //
    Ok(())
}

#[rustfmt::skip]
#[derive(Default)]
pub struct BenchStr {
    pub name: String,   // name
    pub time: f64,      // seconds
    pub is_cycle: bool, // cycles
    pub time_1k: f64,   // seconds per 1k
    pub size: u64,      // bytes
    pub oh_time: f64,   // seconds
    pub oh_size: u64,   // bytes
}

fn set_size(bench_vec: &mut Vec<BenchStr>, in_file: &str) -> anyhow::Result<()> {
    let mut base_time = 0f64;
    let mut base_size = 0u64;
    let re_1 = regex::Regex::new(r"^ *(\d+)\t.*\t([^ ]+)$").unwrap();
    let reader = std::io::BufReader::new(
        std::fs::File::open(in_file)
            .with_context(|| format!("could not open file `{}`", in_file))?,
    );
    for line in reader.lines() {
        let line = line?;
        if let Some(caps) = re_1.captures(&line) {
            //  934281	  26312	    736	 961329	  eab31	cmp_structopt-curl
            let size_s = &caps[1];
            let name_s = &caps[2];
            let name = if name_s.ends_with("-curl") {
                &name_s[0..(name_s.len() - 5)]
            } else {
                name_s
            };
            let i = match bench_vec.iter().position(|x| x.name == name) {
                Some(i) => i,
                None => {
                    let msg = format!("can not find size: {}", name);
                    return Err(anyhow::Error::msg(msg));
                }
            };
            bench_vec[i].size = size_s.parse::<u64>()?;
            if name == "cmp_null_void" {
                base_time = bench_vec[i].time;
                base_size = bench_vec[i].size;
            }
        }
    }
    //
    for bench in bench_vec {
        bench.oh_time = bench.time - base_time;
        bench.oh_size = bench.size - base_size;
    }
    //
    Ok(())
}

fn get_bench(in_file: &str) -> anyhow::Result<Vec<BenchStr>> {
    let mut vec_benchstr: Vec<BenchStr> = Vec::new();
    //
    let re_1 =
        regex::Regex::new(r"^([^ ]+) +time: +[\[][^ ]+ [^ ]+ ([^ ]+) ([^ ]+) [^ ]+ [^ ]+[\]]$")
            .unwrap();
    //
    let reader = std::io::BufReader::new(
        std::fs::File::open(in_file)
            .with_context(|| format!("could not open file `{}`", in_file))?,
    );
    for line in reader.lines() {
        let line = line?;
        if let Some(caps) = re_1.captures(&line) {
            // cmp_structopt::curl::   time:   [302.50 us 302.87 us 303.34 us]
            // cmp_structopt::curl::   time:   [714991.6559 cycles 715483.2743 cycles 716029.3928 cycles]
            let nm = normalize_name(&caps[1])?;
            let tm = normalize_time(&caps[2], &caps[3])?;
            let is_cycle = if &caps[3] == "cycles" { true } else { false };
            let time_1k = if nm.ends_with("^01k") {
                tm
            } else if nm.ends_with("^08k") {
                tm / 8.0
            } else if nm.ends_with("^90k") {
                tm / 90.0
            } else {
                0.0
            };
            //
            vec_benchstr.push(BenchStr {
                name: nm,
                time: tm,
                is_cycle: is_cycle,
                time_1k: time_1k,
                ..BenchStr::default()
            });
        }
    }
    //vec_benchstr.sort_by(|a, b| a.time.partial_cmp(&b.time).unwrap());
    vec_benchstr.sort_by(|a, b| a.name.partial_cmp(&b.name).unwrap());
    //
    Ok(vec_benchstr)
}

fn normalize_name(name_s: &str) -> anyhow::Result<String> {
    Ok(name_s.to_string())
}

fn normalize_time(num_s: &str, unit_s: &str) -> anyhow::Result<f64> {
    let num: f64 = num_s.parse::<f64>()?;
    let unit: f64 = match unit_s {
        "ms" => 0.001,
        "us" => 0.000001,
        "ns" => 0.000000001,
        "ps" => 0.000000000001,
        "cycles" => 1.0,
        _ => {
            let msg = format!("can not convert unit: {}", unit_s);
            return Err(anyhow::Error::msg(msg));
        }
    };
    Ok(num * unit)
}


benches/bench-trait-obj.rs
use criterion::{criterion_group, criterion_main, Criterion};
use criterion_cycles_per_byte::CyclesPerByte;

fn process_one(
    count: i32,
) -> anyhow::Result<(
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
    (&'static str, i32),
)> {
    cmp_polymorphism::do_trait_obj(count)
}

fn criterion_benchmark(c: &mut Criterion<CyclesPerByte>) {
    match process_one(criterion::black_box(criterion::black_box(10))) {
        Ok(((ct, cn), (dt, dn), (ut, un), (rt, rn), (gt, gn))) => {
            assert_eq!(ct, "Meow!");
            assert_eq!(cn, 10);
            assert_eq!(dt, "Woof!");
            assert_eq!(dn, 10 + 10 + 1);
            assert_eq!(ut, "Quack!");
            assert_eq!(un, 10 + 10 + 1 + 10 + 2);
            assert_eq!(rt, "Caw!");
            assert_eq!(rn, 10 + 10 + 1 + 10 + 2 + 10 + 3);
            assert_eq!(gt, "Croak!");
            assert_eq!(gn, 10 + 10 + 1 + 10 + 2 + 10 + 3 + 10 + 4);
        }
        Err(err) => {
            eprintln!("{}", err);
            unreachable!();
        }
    }
    //
    c.bench_function("trait_obj^", |b| {
        b.iter(|| {
            let _r = process_one(criterion::black_box(10));
        })
    });
    //
    {
        let vec = cmp_polymorphism::create_trait_objs(criterion::black_box(1000));
        c.bench_function("trait_obj^vec^id^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^sum^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^rem^01k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_trait_objs(&vec);
            })
        });
    }
    //
    {
        let vec = cmp_polymorphism::create_trait_objs(criterion::black_box(8000));
        c.bench_function("trait_obj^vec^id^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^sum^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^rem^08k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_trait_objs(&vec);
            })
        });
    }
    //
    {
        let vec = cmp_polymorphism::create_trait_objs(criterion::black_box(90000));
        c.bench_function("trait_obj^vec^id^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_id_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^sum^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_sum_trait_objs(&vec);
            })
        });
        c.bench_function("trait_obj^vec^rem^90k", |b| {
            b.iter(|| {
                let _r = cmp_polymorphism::sum_rem_trait_objs(&vec);
            })
        });
    }
}

criterion_group!(
    name = benches;
    config = Criterion::default().with_measurement(CyclesPerByte)
        .warm_up_time(std::time::Duration::from_millis(300))
        .measurement_time(std::time::Duration::from_millis(1500));
    targets = criterion_benchmark);
criterion_main!(benches);
